src/osp/Resource/Resource.h

```diff
@@ -24,29 +24,37 @@
  */
 #pragma once
 
+#include <assert.h>
 #include <map>
+#include <optional>
 #include <string>
 
 namespace osp
 {
 
+struct construct_tag {};
+struct reserve_tag {};
+
 template <class TYPE_T>
 struct Resource
 {
-    Resource(bool loaded)
+    Resource(reserve_tag)
      : m_data()
-     , m_loaded(loaded)
     { }
-    Resource(bool loaded, TYPE_T&& data)
+    Resource(construct_tag)
+     : m_data(TYPE_T())
+    { }
+    Resource(construct_tag, TYPE_T&& data)
      : m_data(std::move(data))
-     , m_loaded(loaded)
     { }
     Resource(Resource&& move) = default;
     Resource(const Resource& copy) = delete;
+    ~Resource()
+    {
+        assert(m_refCount == 0);
```

Comment:
I forgot I put this here. It shouldn't cause any issues if DependRes is doing it's job properly

Reply:
I don't understand what this reference count is doing, if nothing in the Resource class modifies the reference count?

Reply:
The ref count is modified by DependRes. This assert is making sure that there is no references to the resource when it's destructed.